src/algorithms/pid/MergeParticleID.cc

```diff
@@ -0,0 +1,199 @@
+    for(size_t idx_pid = 0; idx_pid < in_pid_collection->size(); idx_pid++) {
+
+      // make the index pair
+      const auto& in_pid = in_pid_collection->at(idx_pid);
+      auto& charged_particle_track_segment = in_pid.getChargedParticle();
+      if(!charged_particle_track_segment.isAvailable()) {
+        m_log->error("PID object found with no charged particle");
+        continue;
+      }
+      auto id_particle = charged_particle_track_segment.id();
+      auto idx_paired  = std::make_pair(idx_coll, idx_pid);
+      m_log->trace("  idx_pid={}  id_particle={}", idx_pid, id_particle);
+
+      // insert in `particle_pids`
+      auto it = particle_pids.find(id_particle);
+      if(it == particle_pids.end())
+        particle_pids.insert({id_particle, {idx_paired}});
+      else
+        it->second.push_back(idx_paired);
+    }
+  }
+
+  // trace logging
+  if(m_log->level() <= spdlog::level::trace) {
+    m_log->trace("{:-<70}","Resulting `particle_pids` ");
+    for(auto& [id_particle, idx_paired_list] : particle_pids) {
+      m_log->trace("id_particle={}", id_particle);
+      for(auto& [idx_coll, idx_pid] : idx_paired_list)
+        m_log->trace("  (idx_coll, idx_pid) = ({}, {})", idx_coll, idx_pid);
+    }
+  }
+
+  // --------------------------------------------------------------------------------
+
+
+  // loop over charged particles, combine weights from the associated `CherenkovParticleID` objects
+  // and create a merged output `CherenkovParticleID` object
+  m_log->trace("{:-<70}","Begin Merging PID Objects ");
+  for(auto& [id_particle, idx_paired_list] : particle_pids) {
+
+    // trace logging
+    m_log->trace("Charged Particle:");
+    m_log->trace("  id = {}", id_particle);
+    m_log->trace("  PID Hypotheses:");
+
+    // create mutable output `CherenkovParticleID` object `out_pid`
+    auto out_pid = out_pids->create();
+    out_pid.setNpe(0.0);
+    out_pid.setRefractiveIndex(0.0);
+    out_pid.setPhotonEnergy(0.0);
+
+    // define `pdg_2_out_hyp`: map of PDG => merged output hypothesis
+    std::unordered_map< decltype(edm4eic::CherenkovParticleIDHypothesis::PDG), edm4eic::CherenkovParticleIDHypothesis > pdg_2_out_hyp;
+
+    // merge each input `CherenkovParticleID` objects associated with this charged particle
+    for(auto& [idx_coll, idx_pid] : idx_paired_list) {
+      const auto& in_pid = in_pid_collections_list.at(idx_coll)->at(idx_pid);
+
+      // logging
+      m_log->trace("    Hypotheses for PID result (idx_coll, idx_pid) = ({}, {}):", idx_coll, idx_pid);
+      Tools::PrintHypothesisTableHead(m_log,6);
+
+      // merge scalar members
+      out_pid.setNpe(out_pid.getNpe() + in_pid.getNpe()); // sum
+      out_pid.setRefractiveIndex( out_pid.getRefractiveIndex() + in_pid.getNpe() * in_pid.getRefractiveIndex() ); // NPE-weighted average
+      out_pid.setPhotonEnergy(    out_pid.getPhotonEnergy()    + in_pid.getNpe() * in_pid.getPhotonEnergy()    ); // NPE-weighted average
```

Comment:
I would not use a structure as accumulator, as it tracks invalid values for RefractiveIndex and PhotonEnergy until they weights are divided out.

Reply:
Do you want the validity of `in_pid` components checked beforehand?

Reply:
No. Just use external variables, create an object once you have all the information.

Reply:
I mean
```
double sum_npe = 0., sum_npe_times_refractive_index = 0., sum_npe_times_photon_energy = 0.;
for () {
  sum_npe += in.pid.getNpe();
  // ...
}
out_pid.setNpe(sum_npe);
// ...

Reply:
2b9872e1